editor/utils/with-history/index.js

```diff
@@ -63,8 +77,18 @@ export default function withHistory( reducer, options = {} ) {
 			return state;
 		}
 
+		let nextPast = [ ...past, present ];
+
+		shouldCreateUndoLevel = ! past.length || shouldCreateUndoLevel;
+
+		if ( ! shouldCreateUndoLevel && shouldOverwriteState( action, previousAction ) ) {
+			nextPast = past;
```

Comment:
The idea with the inverse condition in https://github.com/WordPress/gutenberg/pull/4956#discussion_r167971118 is that it avoids wastefully calculating a `nextPast` above if we don't intend to use it anyways.

```js
let nextPast = past;
if ( shouldCreateUndoLevel || ! shouldOverwriteState( action, previousAction ) ) {
	nextPast = [ ...past, present ];
}
```

Reply:
I hadn't thought about that! Thanks for this tip!